xbmc/rendering/dx/RenderSystemDX.cpp

```diff
@@ -1750,4 +1779,177 @@ void CRenderSystemDX::SetMaximumFrameLatency(uint8_t latency) const
   }
 }
 
+void CRenderSystemDX::UninitHooks()
+{
+  // uninstall
+  LhUninstallAllHooks();
+  // we need to wait for memory release
+  LhWaitForPendingRemovals();
+  SAFE_DELETE(m_hHook);
+  if (m_hDriverModule)
+  {
+    FreeLibrary(m_hDriverModule);
+    m_hDriverModule = nullptr;
+  }
+}
+
+void CRenderSystemDX::InitHooks()
+{
+  DXGI_OUTPUT_DESC outputDesc;
+  if (!m_pOutput || FAILED(m_pOutput->GetDesc(&outputDesc)))
+    return;
+
+  DISPLAY_DEVICEW displayDevice;
+  displayDevice.cb = sizeof(DISPLAY_DEVICEW);
+  DWORD adapter = 0;
+  bool deviceFound = false;
+
+  // delete exiting hooks.
+  UninitHooks();
+
+  // enum devices to find matched
+  while (EnumDisplayDevicesW(NULL, adapter, &displayDevice, 0))
+  {
+    if (wcscmp(displayDevice.DeviceName, outputDesc.DeviceName) == 0)
+    {
+      deviceFound = true;
+      break;
+    }
+    adapter++;
+  }
+  if (!deviceFound)
+    return;
+
+  CLog::Log(LOGDEBUG, __FUNCTION__": Hookind into UserModeDriver on device %S. ", displayDevice.DeviceKey);
+  wchar_t* keyName =
+#ifndef _M_X64
+    // on x64 system and x32 build use UserModeDriverNameWow key
+    CSysInfo::GetKernelBitness() == 64 ? keyName = L"UserModeDriverNameWow" :
+#endif // !_WIN64
+    L"UserModeDriverName";
+
+  DWORD dwType = REG_MULTI_SZ;
+  HKEY hKey = 0;
+  wchar_t value[1024];
+  DWORD valueLength = sizeof(value);
+  LSTATUS lstat;
+
+  // to void \Registry\Machine at the beginning, we use shifted pointer at 18
+  if (ERROR_SUCCESS == (lstat = RegOpenKeyExW(HKEY_LOCAL_MACHINE, displayDevice.DeviceKey + 18, 0, KEY_READ, &hKey))
+   && ERROR_SUCCESS == (lstat = RegQueryValueExW(hKey, keyName, nullptr, &dwType, (LPBYTE)&value, &valueLength)))
+  {
+    // 1. registry value has a list of drivers for each API with the following format: dx9\0dx10\0dx11\0dx12\0\0
+    // 2. we split the value by \0
+    std::vector<std::wstring> drivers;
+    const wchar_t* pValue = value;
+    while (*pValue)
+    {
+      drivers.push_back(std::wstring(pValue));
+      pValue += drivers.back().size() + 1;
+    }
+    // no entries in the registry
+    if (drivers.empty())
+      return;
+    // 3. we take only first three values (dx12 driver isn't needed if it exists ofc)
+    if (drivers.size() > 3)
+      drivers = std::vector<std::wstring>(drivers.begin(), drivers.begin() + 3);
+    // 4. and then reverse vector to start iterate with the best candidate for d3d11 driver
+    std::reverse(drivers.begin(), drivers.end());
+
+    for (auto it = drivers.begin(); it != drivers.end(); ++it)
```

Comment:
instead of `std::reverse` above you can use `for (auto it = drivers.rbegin(); it != drivers.rend(); ++it)` to iterate in reverse order.